src/Codegen/FieldResolver.hack

```diff
@@ -0,0 +1,73 @@
+namespace Slack\GraphQL\Codegen;
+
+use namespace HH\Lib\{C, Dict};
+use namespace Facebook\DefinitionFinder;
+
+
+/**
+ * Utility class which handles traversing the object graph and finding all fields for a class.
+ */
+final class FieldResolver {
+    private dict<string, DefinitionFinder\ScannedClassish> $scanned_classes;
+    private dict<string, dict<string, Field>> $resolved_fields = dict[];
+
+    public function __construct(vec<DefinitionFinder\ScannedClassish> $classes) {
+        $this->scanned_classes = Dict\from_values($classes, $class ==> $class->getName());
+    }
+
+    public function resolveFields(): dict<string, vec<Field>> {
+        foreach ($this->scanned_classes as $class) {
+            if (!$this->shouldResolve($class)) continue;
+            $this->resolveClass($class);
+        }
+        return Dict\map($this->resolved_fields, $fields ==> vec($fields));
+    }
+
+    private function shouldResolve(DefinitionFinder\ScannedClassish $class): bool {
+        $rc = new \ReflectionClass($class->getName());
+        return (
+            $rc->getAttributeClass(\Slack\GraphQL\InterfaceType::class)
+                ?? $rc->getAttributeClass(\Slack\GraphQL\ObjectType::class)
+        ) is nonnull;
+    }
+
+    private function resolveClass(DefinitionFinder\ScannedClassish $class): dict<string, Field> {
+        if (C\contains_key($this->resolved_fields, $class->getName())) {
+            return $this->resolved_fields[$class->getName()];
+        }
+
+        $fields = dict[];
+
+        $parents = $class->getInterfaceNames();
+        $parent_class_name = $class->getParentClassName();
+        if ($parent_class_name is nonnull) {
+            $parents[] = $parent_class_name;
+        }
+
+        foreach ($parents as $parent) {
+            $parent_class = $this->scanned_classes[$parent] ?? null;
+            if ($parent_class) {
+                $fields = Dict\merge($fields, $this->resolveClass($parent_class));
```

Comment:
At some point we'll probably want some extra validation here, instead of just blindly merging the dicts. For example:

- a method with the same name should not have `<<GraphQL\Field(...)>>` with different field names across the hierarchy
- arguments should have the same names (since GraphQL has named arguments, unlike Hack)

A simpler option could be to just enforce that `<<GraphQL\Field(...)>>` can only be used once for each method name across the hierarchy -- e.g. if there's `<<GraphQL\Field('foo')>>` on an interface, adding `<<GraphQL\Field>>` on the overridden method in the subclass would be illegal.

Reply:
> A simpler option could be to just enforce that <<GraphQL\Field(...)>> can only be used once for each method name across the hierarchy -- e.g. if there's <<GraphQL\Field('foo')>> on an interface, adding <<GraphQL\Field>> on the overridden method in the subclass would be illegal.

This seems effectively the same as what you detailed above, and therefore preferable. I'll file an issue.